aips/aip-x.md

```diff
@@ -0,0 +1,237 @@
+- Smart contract needs `signer` to move items into/away from account.
+
+This results in a huge security concern:
+
+- Different smart contract uses the same `signer`  to authenticate an operation. This means that a malicious smart contract can pass the `signer`  to another module and perform the operation on user’s behalf without being acknowledged by the user.
+- `signer`  could also be abused to inflate your account.
+
+We need to design a solution such that:
+
+- Move modules should be able to
+    - define permissions
+    - Framework/transaction can set those permissions
+- Transactions that tries to execute operations that are not included in the list needs to be aborted and rejected by the Aptos VM.
+    - Work with our existing signer module so that users can still set constraints on what transactions invoking existing on chain move functions could invoke.
+
+
+## Impact
+
+This would be a full compatible change to the framework. Existing on chain move modules wouldn't be impacted. We will add those constraints to the Aptos Framework around popular asset types, i.e: Coin/FungibleAsset/Object/NFT.
+
+## Alternative Solutions
+
+Another way to enforce permission is to leverage AIP-56 the runtime access control engine in Move. However, the runtime access control is more suitable to provide coarse grained permissions around modules but it's hard to represent permissions such as withdrawing X APT from a signer. However, we are open to encoperate the changes to the Move on Aptos language extensions in the future if we want to extend the permission system to the general public.
+
+
+## Specification and Implementation Details
+Currently, a signer is a single wrapper around an address to represent that the signature of the transaction passes the address authentication logic. We would like to extend the notion of existing signers in the Move VM so that we could provide APIs to manage contextual data  embedded in such extended signer.
+
+### Lifecycle of a permissioned handle
+
+The lifecycle of the new signer will look like the following:
+
+```jsx
+module aptos_framework::permissioned_signer {
+    struct PermissionedHandle {
+        master_addr: address,
+        permission_addr: address,
+    }
+
+    public fun create_permissioned_handle(
+        master: &signer,
+    ): PermissionedHandle;
+
+    public fun signer_from_permissioned(p: &PermissionedHandle): signer;
+
+    public fun destroy_permissioned_handle(p: PermissionedHandle);
+
+    public fun is_permissioned_signer(s: &signer): bool;
+}
+```
+
+We would introduce a new struct called `PermissionedHandle` that is derived from the original signer. The handle is generated by `create_permissioned_handle_with_expiration`  API, which takes an original signer. In that handle, we will generate a fresh address using auid API to store the permission information that would be needed by the management API.
```

Comment:
code has create_permissioned_handle, while you use create_permissioned_handle_with_expiration here 

we should probably have `create_permissioned_handle` which only creates signer within the current transaction, and create_permissioned_handle_with_expiration which creates it for longer (but upper bounded, and by default short) time window

We need to make sure that default mode - which creates signer valid only within the current transaction, doesn't leak to other transactions in a block, which have same onchain timestamp (as timestamp gets changed only once per block)